src/AspNet.Security.OAuth.SuperOffice/SuperOfficeAuthenticationHandler.cs

```diff
@@ -0,0 +1,163 @@
+﻿/*
+ * Licensed under the Apache License, Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
+ * See https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers
+ * for more information concerning the license and the contributors participating to this project.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IdentityModel.Tokens.Jwt;
+using System.Linq;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Security.Claims;
+using System.Text.Encodings.Web;
+using System.Text.Json;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.OAuth;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace AspNet.Security.OAuth.SuperOffice
+{
+    /// <summary>
+    /// Defines a handler for authentication using SuperOffice.
+    /// </summary>
+    public class SuperOfficeAuthenticationHandler : OAuthHandler<SuperOfficeAuthenticationOptions>
+    {
+        private readonly JwtSecurityTokenHandler _tokenHandler;
+
+        public SuperOfficeAuthenticationHandler(
+            IOptionsMonitor<SuperOfficeAuthenticationOptions> options,
+            ILoggerFactory logger,
+            UrlEncoder encoder,
+            ISystemClock clock,
+            JwtSecurityTokenHandler tokenHandler)
+            : base(options, logger, encoder, clock)
+        {
+            _tokenHandler = tokenHandler;
+        }
+
+        /// <summary>
+        /// The handler calls methods on the events which give the application control at certain points where processing is occurring.
+        /// If it is not provided a default instance is supplied which does nothing when the methods are called.
+        /// </summary>
+        protected new SuperOfficeAuthenticationEvents Events
+        {
+            get { return (SuperOfficeAuthenticationEvents)base.Events; }
+            set { base.Events = value; }
+        }
+
+        /// <inheritdoc />
+        protected override Task<object> CreateEventsAsync() => Task.FromResult<object>(new SuperOfficeAuthenticationEvents());
+
+        /// <inheritdoc />
+        protected override async Task<AuthenticationTicket> CreateTicketAsync(ClaimsIdentity identity, AuthenticationProperties properties, OAuthTokenResponse tokens)
+        {
+            // the access token is a reference token prefixed with the tenant context identifier
+            var contextId = tokens.AccessToken[3..tokens.AccessToken.IndexOf('.', StringComparison.InvariantCultureIgnoreCase)];
```

Comment:
This makes assumptions about indexes that wouldn't give useful exception messages if they occured (e.g. why `3`?)